test/e2e/framework/framework.go

```diff
@@ -0,0 +1,327 @@
+}
+
+// InstallBoskos begins the boskos server and makes sure it is ready to serve
+// before returning the port it is serving on.
+func InstallBoskos(t *testing.T, opt BoskosOptions) string {
+	t.Helper()
+	return runAccessory(t, "boskos", flags(map[string]string{
+		"config":    opt.ConfigPath,
+		"in_memory": "true",
+		"log-level": "debug",
+	})...)
+}
+
+// ConfigResolverOptions are options for running the config server
+type ConfigResolverOptions struct {
+	ConfigPath     string
+	RegistryPath   string
+	ProwConfigPath string
+	FlatRegistry   bool
+}
+
+// InstallConfigResolver begins the configresolver server and makes sure it is ready
+// to serve before returning the port it is serving on.
+func InstallConfigResolver(t *testing.T, opt ConfigResolverOptions) string {
+	t.Helper()
+	return runAccessory(t, "ci-operator-configresolver", flags(map[string]string{
+		"config":        opt.ConfigPath,
+		"registry":      opt.RegistryPath,
+		"prow-config":   opt.ProwConfigPath,
+		"flat-registry": strconv.FormatBool(opt.FlatRegistry),
+		"log-level":     "debug",
+		"cycle":         "2m",
+	})...)
+}
+
+func runAccessory(t *testing.T, command string, args ...string) string {
+	t.Helper()
+	if _, err := exec.LookPath(command); err != nil {
+		t.Fatalf("no `%s` binary was found: %v", command, err)
+	}
+
+	tmpDir := t.TempDir()
+	ctx, cancel := context.WithCancel(context.Background())
+	main, health := getFreePort(t), getFreePort(t)
+	cmd := exec.CommandContext(ctx, command, append(args, flags(map[string]string{
+		"port":        main,
+		"health-port": health,
+	})...)...)
+	log, err := os.Create(filepath.Join(tmpDir, fmt.Sprintf("%s.log", command)))
+	if err != nil {
+		t.Fatalf("could not create log file: %v", err)
+	}
+	cmd.Stdout = log
+	cmd.Stderr = log
+	go func() {
+		if err := cmd.Run(); err != nil && ctx.Err() != context.Canceled {
+			data, readErr := ioutil.ReadAll(log)
+			if readErr != nil {
+				t.Logf("could not read `%s` log: %v", command, readErr)
+			}
+			t.Errorf("`%s` failed: %v\nlogs:%v", command, err, string(data))
+		}
+	}()
+	t.Cleanup(func() {
+		cancel()
+	})
+	if err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (done bool, err error) {
+		resp, err := http.Get(fmt.Sprintf("127.0.0.1:%s/healthz/ready", health))
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				t.Errorf("could not close response body: %v", err)
+			}
+		}()
+		return (resp != nil && resp.StatusCode == http.StatusOK) && err != nil, nil
+	}); err != nil {
+		t.Fatalf("could not wait for `%s` to be healthy: %v", command, err)
+	}
+	return main
+}
+
+func flags(data map[string]string) []string {
+	var out []string
+	for key, value := range data {
+		out = append(out, flag(key, value))
+	}
+	return out
+}
+
+func flag(flag, value string) string {
+	return fmt.Sprintf("--%s=%s", flag, value)
+}
+
+// getFreePort asks the kernel for a free open port that is ready to use.
+func getFreePort(t *testing.T) string {
```

Comment:
I think this needs a lock for threadsafety

Reply:
Why? The `ResolveTPCAddr` should lock implicitly through the kernel, no?

Reply:
sure but there is a timegap between finding a port and the component that wants to use it stating to use it. In that time, we can re-allocate it to something else.

I think we might have to track the ports we handed out?

Reply:
That's a good point, we will need to track them and ask for another if we get a repeat

Reply:
Nothing actually uses this yet (decided to scope this PR to only simple.sh) so I will add this in the next one